test/module/shared_model/validators/field_validator_test.cpp

```diff
@@ -0,0 +1,747 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <limits>
+#include <memory>
+#include <unordered_set>
+
+#include <google/protobuf/descriptor.h>
+#include <google/protobuf/dynamic_message.h>
+#include <boost/format.hpp>
+
+#include "builders/protobuf/queries.hpp"
+#include "builders/protobuf/transaction.hpp"
+#include "module/shared_model/validators/validators_fixture.hpp"
+#include "utils/lazy_initializer.hpp"
+#include "validators/field_validator.hpp"
+
+using namespace shared_model;
+
+class FieldValidatorTest : public ValidatorsTest {
+  validation::FieldValidator field_validator;
+
+ protected:
+  // Function which performs validation
+  using ValidationFunction = std::function<validation::ReasonsGroupType()>;
+  // Function which initializes field, allows to have one type when dealing
+  // with various types of fields
+  using InitFieldFunction = std::function<void()>;
+
+  /**
+   * FieldTestCase is a struct that represents one value of some field,
+   * and expected validation result.
+   */
+  struct FieldTestCase {
+    std::string name;
+    InitFieldFunction init_func;
+    bool value_is_valid;
+    std::string expected_message;
+  };
+
+  // Returns string containing field name and test case name for debug output
+  std::string testFailMessage(const std::string &field_name,
+                              const std::string &testcase_name) const {
+    return (boost::format("Field: %s\nTest Case: %s\n") % field_name
+            % testcase_name)
+        .str();
+  }
+
+ public:
+  // To test one field, validation function is required,
+  // which is always the same, and several test cases, which represent
+  // various inputs
+  using FieldTest = std::pair<ValidationFunction, std::vector<FieldTestCase>>;
+
+  // Run all test cases for given field
+  void runTestCases(const google::protobuf::FieldDescriptor *field) {
+    auto field_name = field->name();
+    // skip field, if already tested
```

Comment:
What's the problem if it runs twice?

Reply:
Twice - no issue, but it can run as many times as number of commands that have this field. Even if it is fast, I would consider 10x increase in test time to be unwanted.

Reply:
Looks quite hackish, but let it be. Hopefully that can save time on debugging in future

Reply:
This is hackish, because we are limited by protobuf reflection, if we can get rid of it in the future, this code will hopefully go away.